server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,404 @@
+   * performed.
+   *
+   * @param context
+   *          the server context.
+   * @param initProps
+   *          map of k, v pairs of initial properties.
+   * @return true if props create, false if the node existed and initialization was skipped.
+   */
+  public synchronized static boolean initSysProps(final ServerContext context,
+      final Map<String,String> initProps) {
+    PropCacheId sysPropsId = PropCacheId.forSystem(context.getInstanceID());
+    return createInitialProps(context, sysPropsId, initProps);
+  }
+
+  /**
+   * Create initial properties if they do not exist. If the node exists, initialization will be
+   * skipped.
+   *
+   * @param context
+   *          the system context
+   * @param propCacheId
+   *          a prop id
+   * @param props
+   *          initial properties
+   * @return true if props create, false if the node existed and initialization was skipped.
+   */
+  public static boolean createInitialProps(final ServerContext context,
+      final PropCacheId propCacheId, Map<String,String> props) {
+
+    try {
+      ZooReaderWriter zrw = context.getZooReaderWriter();
+      if (zrw.exists(propCacheId.getPath())) {
+        return false;
+      }
+      VersionedProperties vProps = new VersionedProperties(props);
+      return zrw.putPersistentData(propCacheId.getPath(),
+          context.getVersionedPropertiesCodec().toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException("Interrupted creating node " + propCacheId, ex);
+    } catch (Exception ex) {
+      throw new IllegalStateException("Failed to create node " + propCacheId, ex);
+    }
+  }
+
+  public PropStoreMetrics getMetrics() {
+    return cacheMetrics;
+  }
+
+  // TODO - evaluate returning the props instead of boolean.
+  @Override
+  public boolean create(PropCacheId propCacheId, Map<String,String> props)
+      throws PropStoreException {
+
+    VersionedProperties vProps = new VersionedProperties(props);
+
+    try {
+
+      var path = propCacheId.getPath();
+      if (!zrw.putPersistentData(path, propCodec.toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL)) {
+        return false;
+      }
+
+      Stat stat = zrw.getStatus(path, propStoreWatcher);
+
+      if (stat.getVersion() != vProps.getNextVersion()) {
+        throw new PropStoreException("Invalid data version on create, have: "
+            + vProps.getNextVersion() + " received " + stat.getVersion(),
+            new IllegalStateException());
+      }
+
+      return true;
+
+    } catch (IOException | KeeperException | InterruptedException ex) {
+      throw new PropStoreException("Failed to serialize properties for " + propCacheId, ex);
+    }
+  }
+
+  @Override
+  public VersionedProperties get(final PropCacheId propCacheId) throws PropStoreException {
+    try {
+
+      checkZkConnection(); // if ZK not connected, block, do not just return a cached value.
```

Comment:
What happens if something calls and ZK is ok and the method returns, but after returning ZK becomes not ok before calling the following two methods?

Reply:
If the value is cached, it will be returned.  If it has not been previously seen, the ZooKeeper read from the caffeine loader will fail.

There is a watcher set that listens for connection events that will clear the cache on disconnection events so the time-window for returning a cached value should be narrow.  I don't think there is a way to avoid some window where a value can be read and returned and simultaneously a ZooKeeper disconnection event because the events are always asynchronous - so I tried to keep that window as small as possible, as well as have the disconnection event eventually recognized asynchronously.

Reply:
What is the benefit of the check?